x/pocketcore/types/session.go

```diff
@@ -124,17 +139,22 @@ func NewSessionNodes(sessionCtx, ctx sdk.Ctx, keeper PosKeeper, chain string, se
 		sessionKey = Hash(sessionKey)
 		// get the node from the array
 		n := nodesAddrs[index.Int64()]
-		//if we already have seen this address we continue as it's either on the list or discarded
+		// if we already have seen this address we continue as it's either on the list or discarded
 		if _, ok := m[n.String()]; ok {
 			continue
 		}
-		//add the node address to the map
+		// add the node address to the map
 		m[n.String()] = struct{}{}
 
 		// cross check the node from the `new` or `end` world state
 		node = keeper.Validator(ctx, n)
-		// if not found or jailed, don't add to session and continue
-		if node == nil || node.IsJailed() || !NodeHasChain(chain, node) || sessionNodes.Contains(node.GetAddress()) {
+		lenNodeChains := int64(len(node.GetChains()))
+		// if not found or jailed or is overstaked to chains
+		if node == nil ||
+			isEnforceMaxChains && lenNodeChains > keeper.MaxChains(ctx) ||
```

Comment:
I'm not 100% sure but I think this could result in a bug. Better safe than sorry

```suggestion
			(isEnforceMaxChains && lenNodeChains > keeper.MaxChains(ctx)) ||
```